exec/java-exec/src/main/java/org/apache/drill/exec/work/WorkManager.java

```diff
@@ -158,38 +165,49 @@ public DrillbitContext getContext() {
     return dContext;
   }
 
-  private ExtendedLatch exitLatch = null; // used to wait to exit when things are still running
-
   /**
    * Waits until it is safe to exit. Blocks until all currently running fragments have completed.
-   *
-   * <p>This is intended to be used by {@link org.apache.drill.exec.server.Drillbit#close()}.</p>
+   * This is intended to be used by {@link org.apache.drill.exec.server.Drillbit#close()}.
    */
   public void waitToExit() {
-    synchronized(this) {
-      if (queries.isEmpty() && runningFragments.isEmpty()) {
-        return;
+    final long startTime = System.currentTimeMillis();
+    final long endTime = startTime + EXIT_TIMEOUT;
+
+    exitLock.lock();
+
+    try {
+      long currentTime;
+      while ((currentTime = System.currentTimeMillis()) < endTime) {
```

Comment:
I'd recommend changing `while` condition to `queries` and `runningFragments` not empty check as a primary condition for exiting. Timeout (waitTime >0) is the secondary condition.

Reply:
Done